src/main/scala/io/keen/client/scala/BatchWriterClient.scala

```diff
@@ -0,0 +1,208 @@
+                error("Failed to send queued events")
+                error(s"$ex")
+            }
+          }
+        }, 1, sendInterval.toMillis, TimeUnit.MILLISECONDS)
+
+        Some(tp)
+    }
+  }
+
+  /**
+   * Flush queued events, removing them from the queue as they are successfully sent.
+   */
+  def sendQueuedEvents(): Unit = {
+    val handleMap = eventStore.getHandles(projectId)
+    val handles = ListBuffer.empty[Long]
+    val events = ListBuffer.empty[String]
+
+    // iterate over all of the event handles in the queue, by collection
+    for ((collection, eventHandles) <- handleMap) {
+      // get each event, and its handle, then add it to a buffer so we can group the events
+      // into smaller batches
+      for (handle <- eventHandles) {
+        handles += handle
+        events += eventStore.get(handle)
+      }
+
+      // group handles separately so we can use them to remove events from the queue once they've
+      // been successfully added
+      val handleGroup: List[ListBuffer[Long]] = handles.grouped(batchSize).toList
+
+      // group the events by batch size, then publish them
+      for ((batch, index) <- events.grouped(batchSize).zipWithIndex) {
+        // publish this batch
+        var response = Await.result(
+          addEvents(s"""{"$collection": [${batch.mkString(",")}]}"""),
+          batchTimeout
+        )
+
+        // handle addEvents responses properly
+        response.statusCode match {
+          case 200 | 201 =>
+            info(s"""${response.statusCode} ${response.body} | Sent ${batch.size} queued events""")
+
+            // remove all of the handles for this batch
+            for (handle <- handleGroup(index)) {
+              eventStore.remove(handle)
+            }
+
+            info(s"""Removed ${handleGroup(index).size} events from the queue""")
+
+          // log but DO NOT remove events from queue
+          case _ => error(s"""${response.statusCode} ${response.body} | Failed to send ${batch.size} queued events""")
+        }
+      }
+    }
+  }
+
+  /**
+   * Flush queued events, sending them to Keen IO on a background thread.
+   */
+  def sendQueuedEventsAsync(): Unit = {
+    // use a thread pool for our async thread so we can use daemon threads
+    val tp = Executors.newSingleThreadExecutor(new ClientThreadFactory)
+
+    // send our queued events in a separate thread
+    tp.execute(new Runnable {
+      def run(): Unit = sendQueuedEvents()
+    })
+  }
+
+  /**
+   * Shut down the threadpool for flushing batch writes, before a final flush of
+   * all events remaining in the queue, run on the main thread.
+   */
+  override def shutdown() = {
+    // Shut down the threadpool, if there is one.
+    scheduledThreadPool foreach { pool =>
+      pool.shutdown()
+      val terminated = pool.awaitTermination(shutdownDelay.toMillis, TimeUnit.MILLISECONDS)
+      if (!terminated) { error("Failed to shutdown scheduled thread pool") }
+    }
```

Comment:
Similarly, this kind of makes you wish you could have some better ability to handle if there's an error here, but we're stuck with the `Unit` return inherited from `Client`. Throwing exceptions is not idiomatic Scala. I'd be open to the thought of changing `shutdown` to return a `Try`.